server/src/main/java/org/apache/kafka/server/BrokerFeatures.java

```diff
@@ -0,0 +1,193 @@
+    private static final Logger log = LoggerFactory.getLogger(BrokerFeatures.class);
+
+    private BrokerFeatures(Features<SupportedVersionRange> supportedFeatures) {
+        this.supportedFeatures = supportedFeatures;
+    }
+
+    public static BrokerFeatures createDefault(boolean unstableFeatureVersionsEnabled) {
+        return new BrokerFeatures(defaultSupportedFeatures(unstableFeatureVersionsEnabled));
+    }
+
+    public static Map<String, VersionRange> createDefaultFeatureMap(BrokerFeatures features) {
+        Map<String, SupportedVersionRange> supportedFeatures = features.supportedFeatures.features();
+        Map<String, VersionRange> result = new HashMap<>();
+        supportedFeatures.forEach((key, value) -> result.put(key, VersionRange.of(value.min(), value.max())));
+        return result;
+    }
+
+    public static Features<SupportedVersionRange> defaultSupportedFeatures(boolean unstableFeatureVersionsEnabled) {
+        Map<String, SupportedVersionRange> features = new HashMap<>();
+        features.put(MetadataVersion.FEATURE_NAME,
+                new SupportedVersionRange(
+                        MetadataVersion.MINIMUM_KRAFT_VERSION.featureLevel(),
+                        unstableFeatureVersionsEnabled ? MetadataVersion.latestTesting().featureLevel()
+                                : MetadataVersion.latestProduction().featureLevel()));
+        PRODUCTION_FEATURES.forEach(feature -> {
+            int maxVersion = unstableFeatureVersionsEnabled ? feature.latestTesting() : feature.latestProduction();
+            if (maxVersion > 0) {
+                features.put(feature.featureName(), new SupportedVersionRange(feature.minimumProduction(), (short) maxVersion));
+            }
+        });
+        return Features.supportedFeatures(features);
+    }
+
+    public static BrokerFeatures createEmpty() {
+        return new BrokerFeatures(Features.emptySupportedFeatures());
+    }
+
+    /**
+     * Returns true if any of the provided finalized features are incompatible with the provided
+     * supported features.
+     *
+     * @param supportedFeatures The supported features to be compared
+     * @param finalizedFeatures The finalized features to be compared
+     * @return - True if there are any feature incompatibilities found.
+     * - False otherwise.
+     */
+    public static boolean hasIncompatibleFeatures(Features<SupportedVersionRange> supportedFeatures,
+                                                  Map<String, Short> finalizedFeatures) {
+        return !incompatibleFeatures(supportedFeatures, finalizedFeatures, false).isEmpty();
+    }
+
+    // For testing only
+    public void setSupportedFeatures(Features<SupportedVersionRange> newFeatures) {
+        Map<String, SupportedVersionRange> combined = new HashMap<>(supportedFeatures.features());
+        combined.putAll(newFeatures.features());
+        supportedFeatures = Features.supportedFeatures(combined);
```

Comment:
It is weird to make `supportedFeatures` mutable for testing-only. Could you please add a new helper for testing?
```java
    public static BrokerFeatures createDefault(boolean unstableFeatureVersionsEnabled, Features<SupportedVersionRange> newFeatures) {
        Map<String, SupportedVersionRange> combined = new HashMap<>(defaultSupportedFeatures(unstableFeatureVersionsEnabled).features());
        combined.putAll(newFeatures.features());
        return new BrokerFeatures(Features.supportedFeatures(combined));
    }
```

Reply:
but the `volatile` can't use with `final` keyword, in this case I think we should reserve the `volatile`.

Reply:
>  If we can't make supportedFeatures immutable, Do we still need to use this helper?

Yes, I proposed making supportedFeatures immutable. Are there any production use cases where supportedFeatures would need to be changed?

Reply:
Yes, you are right, I will make it to final.